src/iuse.cpp

```diff
@@ -4725,18 +4755,42 @@ int iuse::hacksaw( player *p, item *it, bool t, const tripoint & )
         p->add_msg_if_player( m_info, _( "You cannot do that while mounted." ) );
         return 0;
     }
-    const cata::optional<tripoint> pnt_ = choose_adjacent( _( "Cut up metal where?" ) );
+    const static std::set<ter_id> allowed_ter_id {
+        t_chainfence_posts,
+        t_window_enhanced,
+        t_window_enhanced_noglass,
+        t_chainfence,
+        t_chaingate_c,
+        t_chaingate_l,
+        t_window_bars_alarm,
+        t_window_bars,
+        t_reb_cage,
+        t_door_bar_c,
+        t_door_bar_locked,
+        t_bars
+    };
+    const static std::set<furn_id> allowed_furn_id {
+        f_rack
+    };
+    const std::function<bool( tripoint )> f = []( tripoint p ) {
+        if( p == g->u.pos() ) {
+            return false;
+        }
+        const ter_id ter = g->m.ter( p );
+        const auto furn = g->m.furn( p );
+
+        const bool is_allowed = ( allowed_ter_id.find( ter ) != allowed_ter_id.end() ) ||
+                                ( allowed_furn_id.find( furn ) != allowed_furn_id.end() ) ;
+        return is_allowed;
+    };
+
+    const cata::optional<tripoint> pnt_ = choose_adjacent_highlight(
+            _( "Cut up metal where?" ), f, false, true );
     if( !pnt_ ) {
+        p->add_msg_if_player( m_info, _( "You can't cut that." ) );
         return 0;
     }
     const tripoint pnt = *pnt_;
-
-    if( pnt == p->pos() ) {
-        add_msg( m_info, _( "Why would you do that?" ) );
-        p->add_msg_if_player( m_info, _( "You're not even chained to a boiler." ) );
```

Comment:
It's a pity we are getting rid of this funny little thing.

Reply:
Exactly, really feel bad about this part of change, sorry...

I just don't know how to leave these messages and still have a good filter.  
May be just not filter out the player position? But it will be highlighted, looks bad.

Reply:
If we really want to keep it (and other similar messages) we can return an additional string from the filter if a point is not selectable (the "reason"), and use the returned string to replace "Invalid direction" in `choose_adjacent_hilight`.

Reply:
Important messages restored!

Removed "skip selection if there is no directions to choose from". Now even if nothing is highlighted, it will still allow the player to make (probably dumb) choice.
But if there is only one valid direction - it will be chosen. No asking. Or else there will be no point in this PR :)

Now it is kind of balanced.